tools/httpbintest/src/lib.rs

```diff
@@ -0,0 +1,1304 @@
+//! let mut requests_complete = 0;
+//! let request_count = test.requests_count();
+//! match http3_conn.poll(&mut conn) {
+//!     Ok((_stream_id, quiche::h3::Event::Finished)) => {
+//!         requests_complete += 1;
+//!         if requests_complete == request_count {
+//!             if !test.assert() {
+//!                 // test validation failed, handle it
+//!             }
+//!         }
+//!     },
+//!     _ => ()
+//! }
+//! # Ok::<(), quiche::h3::Error>(())
+//! ```
+//!
+//! [quiche]: https://github.com/cloudflare/quiche/
+//! [test]: struct.HttpBinTest.html
+//! [`HttpBinTest`]: struct.HttpBinTest.html
+//! [`assert_hdrs()`]: struct.HttpBinTest.html#method.assert_hdrs
+//! [`requests_count()`]: struct.HttpBinTest.html#method.requests_count
+//! [`set_concurrency()`]: struct.HttpBinTest.html#method.set_concurrency
+//! [`add_response_headers()`]: struct.HttpBinTest.html#method.add_response_headers
+//! [`add_response_body()`]: struct.HttpBinTest.html#method.add_response_body
+//! [`assert()`]: struct.HttpBinTest.html#method.assert
+
+#[macro_use]
+extern crate log;
+
+use std::collections::HashMap;
+
+use quiche::h3::Header;
+
+const USER_AGENT: &str = "quiche-http3bin";
+
+// Stores the request, the expected response headers, and the actual response.
+// The assert_hdrs() method is provided for convenience to validate the received
+// headers match the expected headers.
+#[derive(Clone)]
+pub struct HttpBinReq {
+    url: url::Url,
+    hdrs: Vec<Header>,
+    body: Option<Vec<u8>>,
+    expect_resp_hdrs: Option<Vec<Header>>,
+    pub resp_hdrs: Vec<Header>,
+    resp_body: Vec<u8>,
+}
+
+impl HttpBinReq {
+    pub fn new(
+        method: &str, url: &url::Url, body: Option<Vec<u8>>,
+        expect_resp_hdrs: Option<Vec<Header>>,
+    ) -> HttpBinReq {
+        let mut path = String::from(url.path());
+        if let Some(query) = url.query() {
+            path.push('?');
+            path.push_str(query);
+        }
+
+        let mut hdrs = vec![
+            Header::new(":method", method),
+            Header::new(":scheme", url.scheme()),
+            Header::new(":authority", url.host_str().unwrap()),
+            Header::new(":path", &path),
+            Header::new("user-agent", USER_AGENT),
+        ];
+
+        if let Some(body) = &body {
+            hdrs.push(Header::new("content-length", &body.len().to_string()));
+        }
+
+        HttpBinReq {
+            url: url.clone(),
+            hdrs,
+            body,
+            expect_resp_hdrs: expect_resp_hdrs.clone(),
+            resp_hdrs: Vec::new(),
+            resp_body: Vec::new(),
+        }
+    }
+
+    pub fn assert_hdrs(&self) -> bool {
+        let mut ok = true;
+
+        if let Some(expect_hdrs) = &self.expect_resp_hdrs {
+            for hdr in expect_hdrs {
+                if let Some(h) =
+                    self.resp_hdrs.iter().find(|&x| x.name() == hdr.name())
+                {
+                    ok &= hdr.value() == h.value();
```

Comment:
I think we should just use `assert_eq!()` in these functions (not just `assert_hdrs()`), rather than returning a boolean. Also, it probably makes sense to turn them into macros, so it would be easier to debug problems in the tests (the assert means that you get the expected and actual values in a panic message, and the macro means that you get the correct line numbers without needing a stack trace).